### PR TITLE
Team isaac iteration 4

### PR DESCRIPTION
**Pull request by IsaacFehr**
Grouping functionality and various bug fixes

### CODE REVIEW

File: src/main.js [hunk 1]
```diff
@@ -129,21 +128,38 @@ function mouseMoveListener(e) {
  * - existing link: 
  */
 function singleClickEvent(e) {
+  if(mouseUp != 1 || mouseDown != 1){
+    return;
+  }
```
Comment (stezh) [inline src/main.js]: It's better to move this check up to the event listener. Only trigger this even when mouseUp = 1 and mouseDown = 1.

File: src/main.js [hunk 2]
```diff
@@ -326,7 +350,7 @@ function resetState() {
 function selectDraggedObject(e) {
   var selection = d3.select(e.target)
   if (selection.classed("node-rep")) {
-    dragged_object = e.target;
+    dragged_object = e.target.parentNode;
```
Comment (stezh) [inline src/main.js]: so is that all dragged_object will be select the <g>?

File: src/main.js [hunk 3]
```diff
@@ -396,32 +421,47 @@ function drawSelectionArea(e){
...
+    } else{
+      selection_area.attr("width", rectX - e.pageX + width);
+      selection_area.attr("x", e.pageX);
+    }
+    if( e.pageY > midY ){
+      selection_area.attr("height", e.pageY - rectY);
+    } else{
+      selection_area.attr("height", rectY - e.pageY + height);
+      selection_area.attr("y", e.pageY);
+    }
```
Comment (stezh) [inline src/main.js]: Can this be separated into new functions called getSelectionWidth() and getSelectionHeight()? Both of them will return value of width/height and whether it should set x/y.

File: src/main.js (hunk 1 above)
Comment (IsaacFehr) [reply src/main.js]: removed this functionality, and instead reset timers in resetState()

File: src/main.js [hunk 4]
```diff
@@ -129,21 +128,38 @@ function mouseMoveListener(e) {
...
+        var addedNode = addNode();
+        drawNode(addedNode, e.clientX, e.clientY, radius);
+        let new_children_ids = String(e.target.getAttribute("children_ids"))
+          .split(" ")
+          .filter(x => x);
+        new_children_ids.push(addedNode.id);
+        new_children_ids = new_children_ids.join(' ');
+        e.target.setAttribute("children_ids", new_children_ids);
+        break;
+      default:
```
[the remark this answers is not in the document]
Comment (IsaacFehr) [reply src/main.js]: Done.  It is now in addNodeToGroup(node, group)

**Pull request merged**